File: 16_滑动窗口/力扣加加/992. K 个不同整数的子数组.ts
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 * @summary 题目转换成求：不超过k种元素的子数组个数 - 不超过k-1种元素的子数组个数(转化为水果成蓝问题)
 */
const subarraysWithKDistinct = function (nums: number[], k: number): number {
  const notMoreThan = (threshold: number): number => {
    let res = 0
    let left = 0
    const counter = new Map<number, number>()

    for (let right = 0; right < nums.length; right++) {
      const curNum = nums[right]
      counter.set(curNum, (counter.get(curNum) || 0) + 1)

      while (counter.size > threshold) {
        const preNum = nums[left]
        const count = counter.get(preNum)!
        if (count === 1) counter.delete(preNum)
        else counter.set(preNum, count - 1)
        left++
      }

      res += right - left + 1 // 以r结尾的子数组，多少个合理
    }

    return res
  }

  return notMoreThan(k) - notMoreThan(k - 1)
}

console.log(subarraysWithKDistinct([1, 2, 1, 2, 3], 2))

File: 22_专题/atMoskK/992. K 个不同整数的子数组.py
# 给定一个正整数数组 nums和一个整数 k，返回 nums 中 「好子数组」 的数目。
# !如果 nums 的某个子数组中不同整数的个数恰好为 k，则称 nums 的这个连续、不一定不同的子数组为 「好子数组 」。

from collections import defaultdict
from typing import List


class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        if k <= 0:
            return 0

        def cal(ceiling: int) -> int:
            """计算不同整数的个数小于等于 ceiling 的子数组的个数."""
            res, left, n = 0, 0, len(nums)
            counter = defaultdict(int)
            for right in range(n):
                counter[nums[right]] += 1
                while left <= right and len(counter) > ceiling:
                    removed = nums[left]
                    counter[removed] -= 1
                    if counter[removed] == 0:
                        del counter[removed]
                    left += 1
                res += right - left + 1
            return res

        return cal(k) - cal(k - 1)
